pipenv/utils/dependencies.py

```diff
@@ -0,0 +1,393 @@
+        .*                  # Trailing garbage.
+        $                   # End of line.
+    """,
+        re.VERBOSE,
+    )
+
+    match = version_pattern.match(version_line)
+    if not match:
+        return None
+    return match.groupdict(default="0")
+
+
+def python_version(path_to_python):
+    from pipenv.vendor.pythonfinder.utils import get_python_version
+
+    if not path_to_python:
+        return None
+    try:
+        version = get_python_version(path_to_python)
+    except Exception:
+        return None
+    return version
+
+
+def clean_pkg_version(version):
+    """Uses pip to prepare a package version string, from our internal version."""
+    return pep440_version(str(version).replace("==", ""))
+
+
+class HackedPythonVersion:
+    """A Beautiful hack, which allows us to tell pip which version of Python we're using."""
+
+    def __init__(self, python_version, python_path):
+        self.python_version = python_version
+        self.python_path = python_path
+
+    def __enter__(self):
+        # Only inject when the value is valid
+        if self.python_version:
+            os.environ["PIPENV_REQUESTED_PYTHON_VERSION"] = str(self.python_version)
+        if self.python_path:
+            os.environ["PIP_PYTHON_PATH"] = str(self.python_path)
+
+    def __exit__(self, *args):
+        # Restore original Python version information.
+        try:
+            del os.environ["PIPENV_REQUESTED_PYTHON_VERSION"]
+        except KeyError:
+            pass
+
+
+def get_canonical_names(packages):
+    """Canonicalize a list of packages and return a set of canonical names"""
+    from pipenv.vendor.packaging.utils import canonicalize_name
+
+    if not isinstance(packages, Sequence):
+        if not isinstance(packages, str):
+            return packages
+        packages = [packages]
+    return {canonicalize_name(pkg) for pkg in packages if pkg}
+
+
+def is_file(package):
+    """Determine if a package name is for a File dependency."""
+    if hasattr(package, "keys"):
+        return any(key for key in package.keys() if key in ["file", "path"])
+
+    if os.path.exists(str(package)):
+        return True
+
+    for start in SCHEME_LIST:
+        if str(package).startswith(start):
+            return True
+
+    return False
+
+
+def pep440_version(version):
+    """Normalize version to PEP 440 standards"""
+    # Use pip built-in version parser.
+    from pipenv.vendor.pip_shims import shims
+
+    return str(shims.parse_version(version))
+
+
+def pep423_name(name):
+    """Normalize package name to PEP 423 style standard."""
+    name = name.lower()
+    if any(i not in name for i in (VCS_LIST + SCHEME_LIST)):
+        return name.replace("_", "-")
+
+    else:
+        return name
+
+
+def get_vcs_deps(
+    project=None,
+    dev=False,
+    pypi_mirror=None,
+    packages=None,
+    reqs=None
+):
+    from pipenv.vendor.requirementslib.models.requirements import Requirement
```

Comment:
Why sad face here? Can this be moved one line up?

Reply:
Sad face?  Can which be moved up a line?  The blank line 128 or the import statement to the top of the file?  Aside:  I'd like to do another PR soon that introduces pre-commit hooks and linting setup similar to what pip and many other projects have.

Reply:
L136 looks like a sad face. IMHO it could be moved up.

Reply:
Oh, I see it now lol.  I honestly suspect that the flake8 and pre-commit linting auto-formatting stuff I want to add to the code base would prefer this same for any method  whose parameters need to go on multiple lines though it does look like they could all fit on one line here.   The reality is I just moved existing code (https://github.com/pypa/pipenv/pull/4990/files#diff-c0f7c08428b42c6f1d1e153d4286ad792e8e26154e342718fc3e5b9bcbb39453L1883) here in this PR so don't want to get into the business of reformatting it when I have bigger plans to try and apply auto linting to just the pipenv files (and not the patched vendor ones) for that will be a separate PR/issue.

Reply:
ok. I see.